openmc/mesh.py

```diff
@@ -2774,29 +2804,129 @@ def write_data_to_vtk(
 
         writer.Write()
 
+    def _write_data_to_vtk_hdf_format(
+        self,
+        filename: PathLike | None = None,
+        datasets: dict | None = None,
+        volume_normalization: bool = True,
+    ):
+        def append_dataset(dset, array):
+            """Convenience function to append data to an HDF5 dataset"""
+            origLen = dset.shape[0]
+            dset.resize(origLen + array.shape[0], axis=0)
+            dset[origLen:] = array
+
+        if self.library != "moab":
+            raise NotImplemented("VTKHDF output is only supported for MOAB meshes")
+
+        # the self.connectivity contains an arrays of length 8, in the case of
+        # MOAB tetrahedra mesh elements, the last 4 values are -1 and can be removed
+        trimmed_connectivity = []
+        for cell in self.connectivity:
+            # Find the index of the first -1 value, if any
+            first_negative_index = np.where(cell == -1)[0]
+            if first_negative_index.size > 0:
+                # Slice the array up to the first -1 value
+                trimmed_connectivity.append(cell[: first_negative_index[0]])
+            else:
+                # No -1 values, append the whole cell
+                trimmed_connectivity.append(cell)
+        trimmed_connectivity = np.array(trimmed_connectivity, dtype="int32").flatten()
+
+        # MOAB meshes supports tet elements only so we know it has 4 points per cell
```

Comment:
```suggestion
        # MOAB meshes supports tet elements only so we know it has 4 points per cell
```

This leaves us in one of those awkward states where some features work for some meshes but not others. I might look into adding support for hex meshes as well here.

Reply:
I have raised an issue https://github.com/openmc-dev/openmc/issues/3620